static/index.js

```diff
@@ -139,23 +139,43 @@ window.addEventListener('popstate', (event) => {
 // Add event delegation for label links
 document.addEventListener('click', (event) => {
   const target = event.target.closest('a')
-  if (target && target.href) {
-    const url = new URL(target.href, window.location.origin)
-    const labelQuery = url.searchParams.get('q')
-    if (labelQuery !== null) {
-      event.preventDefault()
-      event.stopPropagation()
-
-      const inputEvent = new Event('input', { bubbles: true })
-      const changeEvent = new Event('change', { bubbles: true })
-      input.value = labelQuery
-      input.dispatchEvent(inputEvent)
-      input.dispatchEvent(changeEvent)
-
-      list.scrollUp()
-      list.goSearch(labelQuery.toLowerCase(), 'relevance', 1)
+  if (!target || !target.href) {
+    return
+  }
+
+  const url = new URL(target.href, window.location.origin)
+  const oldQuery = input.value
   const urlParams = new URLSearchParams(window.location.search)
+
+  let newQuery = url.searchParams.get('q')
+  if (newQuery === null) {
+    return
+  }
+
+  event.preventDefault()
+  event.stopPropagation()
+
+  if (target.closest('form')) {
+    // the shortcuts in the form should provide an additional narrowing of search,
+    // not replace it
+    if (oldQuery.includes('label:') && newQuery.includes('label:')) {
+      newQuery = oldQuery.replace(/label:("[^"]+"|\S+)/, newQuery)
+    } else if (oldQuery.includes('platform:') && newQuery.includes('platform:')) {
+      newQuery = oldQuery.replace(/platform:("[^"]+"|\S+)/, newQuery)
+    } else if (oldQuery.includes('author:') && newQuery.includes('author:')) {
+      newQuery = oldQuery.replace(/author:("[^"]+"|\S+)/, newQuery)
+    } else {
+      newQuery = oldQuery + ' ' + newQuery
```

Comment:
Trim `newQuery` assignment here or line 172? Or do we not care about an extra space getting into the query?

Reply:
That is significant. There is a leading space and a redirect then without that space which corrupts the history.  (Just click on the same label over and over.)

Reply:
fixed in 38fb6d4dd73abc0097adeef5cd6ca34b08e0c868